feat:Improve Binder change detection

onuridrisoglu opened the pull request:
## Description

Fixes #19260

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [ ] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [ ] I have added a description following the guideline.
- [ ] The issue is created in the corresponding repository and I have referenced it.
- [ ] I have added tests to ensure my change is effective and works as intended.
- [ ] New and existing tests are passing locally with my change.
- [ ] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

onuridrisoglu committed:
Introduced binder change detection for readBean case

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vaadin/flow?pullRequest=19488) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vaadin/flow?pullRequest=19488) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vaadin/flow?pullRequest=19488) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vaadin/flow?pullRequest=19488) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vaadin/flow?pullRequest=19488) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vaadin/flow?pullRequest=19488) it.</sub>

tepi committed:
Merge branch 'main' into issues/19260-binder-change-detection

tepi commented on flow-data/src/main/java/com/vaadin/flow/data/binder/Binder.java:
```diff
@@ -1929,18 +1931,41 @@ public static <BEAN> Binder<BEAN> withPropertySet(
      *            the binding whose value has been changed
+     * @param event
+     *            the value change event
      */
-    protected void handleFieldValueChange(Binding<BEAN, ?> binding) {
-        changedBindings.add(binding);
-
+    protected void handleFieldValueChange(Binding<BEAN, ?> binding, ValueChangeEvent<?> event) {
```
Please avoid changing this method signature since it would be a breaking change and require waiting for the next major. Instead, one option could be to deprecate the current one, and if that is used, just skip the new change detection logic and function as before.

onuridrisoglu replied:
Added the current `handleFieldValueChange` back as deprecated

tepi commented on flow-data/src/main/java/com/vaadin/flow/data/binder/Binder.java:
```diff
@@ -1929,18 +1931,41 @@ public static <BEAN> Binder<BEAN> withPropertySet(
+     * @return true if the field value is reverted
+     */
+    protected boolean isRevertedToInitialValue(Binding<BEAN, ?> binding, Object newValue) {
+		return Objects.equals(bindingInitialValuesMap.get(binding), newValue);
```
The logic here relies on `equals` of the value type being implemented properly. That is something the Binder has not required before, so it's potentially a breaking change.

tepi replied:
Also, it migh make more sense to add Binding-level API for overriding the logic for checking if value is reverted. That way if custom logic is needed, one does not need to extend Binder to do it.

onuridrisoglu replied:
introduced an equality predicate (I'm open for naming suggestions) that uses `Objects::equals` as default, but it can be overridden in binding-level

tepi commented:
@mshabarov This change will introduce a slight change in behavior of Binder change tracking due to requiring `equals` to be implemented properly by value types. As well as reverting the tracked change if the value is reverted, which is not currently done (that is, the binding is considered as changed anyway). Do you think we'll need an enable/disable switch for this new change tracking?

onuridrisoglu committed:
Added equality predicate to Binding level

and deprecated the old handleFieldValueChange method

knoobie commented on flow-data/src/main/java/com/vaadin/flow/data/binder/Binder.java:
```diff
@@ -949,6 +952,18 @@ BindingBuilder<BEAN, TARGET> asRequired(
+         * @param equalityPredicate the predicate to use for equality comparison
+         * @return this {@code BindingBuilder}, for method chaining
+         */
+        public BindingBuilder<BEAN, TARGET> withEqualityPredicate(SerializableBiPredicate<Object, Object> equalityPredicate);
```
Vote 4 default implementation; otherwise this is a breaking change

onuridrisoglu replied:
Done

onuridrisoglu committed:
Fixed unit test error

onuridrisoglu committed:
Added default implementation for the new withEqualityPredicate method

onuridrisoglu committed:
Added missing line from the deprecated method

mshabarov commented on flow-data/src/main/java/com/vaadin/flow/data/binder/Binder.java:
```diff
@@ -1810,6 +1850,7 @@ void setIdentity() {
 
     private BindingExceptionHandler exceptionHandler = new DefaultBindingExceptionHandler();
 
+    private Map<Binding<BEAN, ?>, Object> bindingInitialValuesMap = new HashMap<>();
```
What if the initial values would be a class field in the `BindingImpl` instead and have a getter method in `Binding`? Then you can call newly added method `Binding::getInitialValue`.

onuridrisoglu replied:
I refactored the changes and removed `bindingInitialValuesMap`. Now the initial value is kept in Binding level which is set during readBean.

mshabarov commented:
> @mshabarov This change will introduce a slight change in behavior of Binder change tracking due to requiring `equals` to be implemented properly by value types. As well as reverting the tracked change if the value is reverted, which is not currently done (that is, the binding is considered as changed anyway). Do you think we'll need an enable/disable switch for this new change tracking?

I'd propose to set the equality predicate `null` by default (or some other value that marks it not defined), that means the change detection feature is not enabled by default. Then if this predicate is set by `BindingBuilder::withEqualityPredicate`, then this feature is enabled. In this way this would be backwards compatible.

onuridrisoglu committed:
Changes for review comments:

- removed bindingInitialValuesMap from Binder, and kept the initial
value in binding level instead
- removed the default implementation for equalityPredicate and now this
feature will only be active if an equalityPredicate is set

onuridrisoglu committed:
Merge branch 'main' into issues/19260-binder-change-detection

onuridrisoglu commented on flow-data/src/main/java/com/vaadin/flow/data/binder/Binder.java:
```diff
@@ -320,6 +321,8 @@ default BindingValidationStatus<TARGET> validate() {
          */
         boolean hasChanges();
+        
+        SerializableBiPredicate<Object, Object> getEqualityPredicate();
```
done

onuridrisoglu commented:
> > @mshabarov This change will introduce a slight change in behavior of Binder change tracking due to requiring `equals` to be implemented properly by value types. As well as reverting the tracked change if the value is reverted, which is not currently done (that is, the binding is considered as changed anyway). Do you think we'll need an enable/disable switch for this new change tracking?
> 
> I'd propose to set the equality predicate `null` by default (or some other value that marks it not defined), that means the change detection feature is not enabled by default. Then if this predicate is set by `BindingBuilder::withEqualityPredicate`, then this feature is enabled. In this way this would be backwards compatible.

I did the changes, however my only concern is now `withEqualityPredicate ` should be called for each binding in order to activate this feature.

onuridrisoglu committed:
fixed format issues

mshabarov committed:
Merge branch 'main' into issues/19260-binder-change-detection

mshabarov reviewed (approved):
Looks good to me, but I haven't tested it.

I agree with the concern about having this API only on a binding level, would be good to allow doing the same for the whole form.

mshabarov commented:
Just discussed with @tepi, the Binder-level API could be done in a way:
1. Adding a method on Binder's level that enables/disables usage of these predicates on bindings level.
2. Then, if enabled, Binder will use a configured predicate for a binding, if it exists, otherwise it uses `equals` as a fallback. If disabled, nothing is used (default).

mshabarov reviewed (changes requested):
Setting to request changes

tepi commented:
> Just discussed with @tepi, the Binder-level API could be done in a way:
> 
>     1. Adding a method on Binder's level that enables/disables usage of these predicates on bindings level.
> 
>     2. Then, if enabled, Binder will use a configured predicate for a binding, if it exists, otherwise it uses `equals` as a fallback. If disabled, nothing is used (default).

@onuridrisoglu fyi, I'll pick this up and do the changes

tepi committed:
Add global setting and default comparison

tepi committed:
Add tests

tepi committed:
Merge branch 'main' into issues/19260-binder-change-detection
